File: First_semester/Homework_7/task_3/task_3/stack.cpp
#include "stdafx.h"
#include "stack.h"
#include <iostream>

struct Stack
{
	StackElement *head = nullptr;
};

struct StackElement
{
	Value value = '\0';
	StackElement *next = nullptr;
};

Stack *createStack()
{
	Stack *stack = new Stack;
	return stack;
}

void push(Value value, Stack *stack)
{
	StackElement *newElement = new StackElement;
	newElement->value = value;
	if (!stack->head)
	{
		stack->head = newElement;
		return;
	}
	newElement->next = stack->head;
	stack->head = newElement;
}

void pop(Stack *stack)
{
	if (!stack->head)
	{
		std::cout << "	Stack is empty" << std::endl;
		return;
	}
	StackElement *temp = stack->head;
	stack->head = stack->head->next;
	delete temp;
}

Value getValue(Stack *stack)
{
	return stack->head->value;
}

bool empty(Stack *stack)
{
	if (!stack->head)
	{
		return true;
	}
	else
	{
		return false;
	}
}

void printStack(Stack *stack)
{
	StackElement *i = stack->head;
	while (i)
	{
		std::cout << i->value << " ";
		i = i->next;
	}
}

void deleteStack(Stack *stack)
{
	if (!stack->head)
	{
		std::cout << "Stack is empty" << std::endl;
		return;
	}

	while (stack->head)
	{
		StackElement *temp = stack->head->next;
		delete stack->head;
		stack->head = temp;
	}
}

File: Second_semester/Homework_2/Stack/Stack/Stack.cs
using System;

namespace StackNamespace
{
    /// <summary>
    /// LIFO
    /// </summary>
    public class Stack : IStack
    {
        public void Push(int value)
        {
            StackElement newElement = new StackElement(value);
            if (this.head == null)
            {
                this.head = newElement;
                return;
            }

            newElement.Next = this.head;
            this.head = newElement;
        }

        public int Top()
        {
            if (this.head == null)
            {
                throw new EmptyStackException("you are trying to access a non-existent object");
            }

            return this.head.Value;
        }

        public void Pop()
        {
            if (this.head == null)
            {
                throw new EmptyStackException("you are trying to access a non-existent object");
            }

            this.head = this.head.Next;
        }

        public bool IsEmpty() => this.head == null;

        public void PrintStack()
        {
            if (this.head == null)
            {
                Console.WriteLine("Stack is empty");
            }

            StackElement i = this.head;
            while (i != null)
            {
                Console.WriteLine(i.Value + " ");
                i = i.Next;
            }
        }

        private class StackElement
        {
            /// <summary>
            /// class constructor
            /// </summary>
            /// <param name="value"></param>
            public StackElement(int value)
            {
                this.Value = value;
            }

            public int Value { get; set; }
            public StackElement Next { get; set; }
        }

        private StackElement head;
    }
}